qiskit_experiments/warnings.py

```diff
@@ -0,0 +1,245 @@
+                message += f"This class will be removed in Qiskit Experiments {version_removed}. "
+            else:
+                message += "This class will be removed in future release. "
+            warnings.warn(message, DeprecationWarning, stacklevel=2)
+            instance = object.__new__(new_cls or deprecated_cls)
+            instance.__init__(*args, **kwargs)
+            return instance
+
+        cls.__new__ = new
+        return cls
+
+    return patch_new
+
+
+def deprecated_options(
+    options_map: Dict[str, str],
+    version_removed: Optional[str] = None,
+) -> Callable:
+    """A set options method decorator to show deprecation warning for deprecated options.
+
+    Args:
+        options_map: A dictionary of old option to new option. If new option value is ``None``
+            it simply removes the option from set options method.
+        version_removed: The Qiskit Experiment version that this class is removed.
+
+    Examples:
+
+        .. code-block::
+
+            class SomeExperiment(BaseExperiment):
+                @deprecated_options(
+                    options_map={"key1": "new_key1", "key2": None},
+                    version_removed="0.3",
+                )
+                def set_experiment_options(self, **fields):
+                    super().set_experiment_options()
+
+    Returns:
+        Deprecated set options method.
+    """
+
+    def update_signature(set_options_method):
+        @functools.wraps(set_options_method)
+        def _wrap(self, **fields):
+            deprecated = options_map.keys() & fields.keys()
+            if any(deprecated):
+                all_options = ", ".join(f"'{dep}'" for dep in deprecated)
+                message = f"Options {all_options} have been deprecated and "
+                if version_removed:
+                    message += f"will be removed in Qiskit Experiments {version_removed}. "
+                else:
+                    message += "will be removed in future release. "
+                clsname = set_options_method.__qualname__.split(".")[0]
+                message += f"If this is a loaded '{clsname}' class instance, "
+                message += (
+                    "please save the experiment again for further retrieval with future software."
+                )
+                warnings.warn(message, DeprecationWarning, stacklevel=2)
+                for dep in deprecated:
+                    new_opt = options_map[dep]
+                    if new_opt is not None:
+                        warnings.warn(
+                            f"Option '{dep}' is now replaced with '{new_opt}'.",
+                            UserWarning,
+                            stacklevel=2,
+                        )
+                        fields[new_opt] = fields[dep]
+                    del fields[dep]
+            set_options_method(self, **fields)
+
+        return _wrap
+
+    return update_signature
+
+
+def deprecated_init_args(
```

Comment:
Does this work with saving Experiment and Analysis configs if its used to patch experiments? Can you add some unit tests to ensure this decorate behaves as expected.

Reply:
Also why is this necessary as as separate method to just having a `deprecate_arguments` function for a generic function (like terra has) that can be applied to a functions `__init__` method?

Reply:
Test is added. The reason why this is a class decorator might be my preference, but this design helps to tell a developer this can be only applied to `__init__` method. Otherwise one may decorate some public method and mess up the instance configuration. In addition, this easily prevents one accidentally patch non-experiment/analysis class constructor, see test in https://github.com/Qiskit/qiskit-experiments/pull/751/commits/e2a00241759dbe3b8640f65cd5e686c348636dbc

Reply:
In that case this doc string should explicitly state that this only works for BaseExperiment and BaseAnalysis. Also this function only appears to deprecate kwargs, not positional args so it should be names that.

I am still quite worried about this working correctly. All the cases in these tests should work: https://github.com/Qiskit/qiskit-experiments/blob/main/test/test_store_init_args.py when wrapped with this decorator.

Another case I don't think this captures is if you are not just renaming an argument, so you want to keep the original arg and use it in the code to be able to have backwards compatibility. Just removing the arg like here with value None is not going to be sufficient in those cases. This would also apply to the above deprecate options one.

Overall I would probably rather remove this function entirely or spend some more time testing it, and just merge the other decorators for deprecation functions and classes for now.

Reply:
That's fine. But all reviewers and developers must manually implement the same warning and logic just in case. Indeed these logics are not important for my current work so it doesn't motivate me to continue.